src/app/loadout-builder/auto-optimizer.ts

```diff
@@ -0,0 +1,434 @@
+    if (!valid) {
+      return LoadoutError.BadSearchQuery;
+    }
+    searchFilter = autoOptContext.filterFactory(loadoutParameters.query);
+  } else {
+    searchFilter = _.stubTrue;
+  }
+
+  const [items, subclass] = getItemsAndSubclassFromLoadout(
+    autoOptContext.itemCreationContext,
+    loadout.items,
+    autoOptContext.store,
+    autoOptContext.allItems
+  );
+
+  const armorEnergyRules: ArmorEnergyRules = {
+    minItemEnergy: MIN_LO_ITEM_ENERGY,
+    assumeArmorMasterwork: loadoutParameters.assumeArmorMasterwork!,
+  };
+
+  const armorItems = items
+    .filter((i) => i.loadoutItem.equip && i.item.bucket.inArmor && i.item.energy)
+    .map((i) => i.item);
+  // A loadout must have 5 equipped armor 2.0 items for comparisons
+  // between LO runs and the current loadout to be meaningful
+  if (armorItems.length !== 5) {
+    return LoadoutError.NotAFullArmorSet;
+  }
+
+  const exotic = armorItems.find((i) => i.isExotic);
+  if (!matchesExoticArmorHash(loadoutParameters.exoticArmorHash, exotic)) {
+    return LoadoutError.DoesNotRespectExotic;
+  }
+  loadoutParameters.exoticArmorHash = exotic?.hash;
+
+  let originalLoadoutMods = getModsFromLoadout(
+    autoOptContext.itemCreationContext.defs,
+    loadout,
+    autoOptContext.unlockedPlugs
+  );
+  // Remove deprecated mods, since the only logical thing is for the user to drop them.
+  originalLoadoutMods = originalLoadoutMods.filter(
+    (mod) => !deprecatedMods.includes(mod.resolvedMod.hash)
+  );
+
+  const { unassignedMods, invalidMods } = fitMostMods({
+    defs: autoOptContext.itemCreationContext.defs,
+    items: armorItems,
+    plannedMods: originalLoadoutMods.map((mod) => mod.resolvedMod),
+    armorEnergyRules,
+  });
+
+  // If our loadout can't even fit the mods it was saved with (except for deprecated mods)
+  // it's an unfair comparison, so bail here.
+  // TODO: Maybe consider all items for mod categorization and ignore invalidMods, since
+  // none of the owned items can fit invalidMods in that case and that doesn't cause a difference between loadouts.
+  if (unassignedMods.length || invalidMods.length) {
+    return LoadoutError.ModsDontFit;
+  }
+
+  const setStats = getLoadoutStats(
+    autoOptContext.itemCreationContext.defs,
+    autoOptContext.classType,
+    subclass,
+    armorItems,
+    originalLoadoutMods.map((mod) => mod.resolvedMod),
+    armorEnergyRules
+  );
+
+  // Save back the actual mods for LO to use
+  loadoutParameters.mods = originalLoadoutMods
+    .filter(
+      (mod) =>
+        // drop artifice mods (always picked automatically per set)
+        mod.resolvedMod.plug.plugCategoryHash !== PlugCategoryHashes.EnhancementsArtifice &&
+        // drop general mods if picked automatically
+        (!loadoutParameters?.autoStatMods ||
+          mod.resolvedMod.plug.plugCategoryHash !== PlugCategoryHashes.EnhancementsV2General)
+    )
+    .map((mod) => mod.originalModHash);
```

Comment:
Maybe we can get away with enabling auto mods if and only if the loadout has any stat mods, regardless the setting?